Arm

sofiebudman opened the pull request.

CoolSpy3 commented on .SysId/sysid.json:
```diff
@@ -0,0 +1,10 @@
+    "mode": "Client (NT4)",
+    "serverTeam": "199"
+  }
+}
```
This file should be deleted

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/Robot.java:
```diff
@@ -28,6 +28,7 @@ public void robotPeriodic() {
     MotorErrors.printSparkMaxErrorMessages();
   }
 
+
```
Delete newline

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/Constants.java:
```diff
@@ -13,12 +15,23 @@
         }
         public static final class Manipulator {
-            public static final int port = 1;
+            public static final int port = 0;
```
Driver should be port `0`, Manipulator should be port `1`

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/Constants.java:
```diff
@@ -13,12 +15,23 @@
+            public static final int intake = Button.kA.value;
+            
+            public static final int cycleUp = 1;//FIXME use correct buttons
+            public static final int cycleDown = 2;
```
Use button constants

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java:
```diff
@@ -4,25 +4,35 @@
-  private void configureButtonBindingsManipulator() {}
+  private void configureButtonBindingsManipulator() {
+    new JoystickButton(manipulatorController, Constants.OI.Manipulator.cycleUp).onTrue(new InstantCommand(()->arm.cycleUp()));
+    new JoystickButton(manipulatorController, Constants.OI.Manipulator.cycleDown).onTrue(new InstantCommand(()->arm.cycleDown()));
```
arm::cycleUp, arm::cycleDown

~~The commands should require Arm~~

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+public class Arm extends SubsystemBase
+{
+  public CANSparkMax motor = MotorControllerFactory.createSparkMax(Constants.Arm.port, TemperatureLimit.NEO);
+  public RelativeEncoder motorLencoder = motor.getEncoder();
```
Make `private final`

CoolSpy3 replied:
Change name to just `motorEncoder`

CoolSpy3 replied:
Use `SparkMaxAbsoluteEncoder`

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+  
+  private double setpoint = 2.2;
+  private double FFvelocity = 6;
+  private double FFaccel = 6;
```
Move to `Constants` and make `public static final`

If you need them to change via SmartDashboard, make the initial values constant and add the word default to the name.

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+    INTAKE(0), MID(1), HIGH(2);
+    
+    public double value; //not static so SmartDashboard can touch [IMPORTANT TO KNOW!]
+    ArmPreset(double value) {
```
Make private

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+  public enum ArmPreset {
+    INTAKE(0), MID(1), HIGH(2);
+    
+    public double value; //not static so SmartDashboard can touch [IMPORTANT TO KNOW!]
```
Add newline after

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+        case HIGH: return MID;
+      }
+      return null;
+    }
```
This requires updating the methods if new values are added/removed. They also shouldn't be able to return `null`.

I recommend `return values()[(ordinal() + 1) % values().length]` and `return values()[(ordinal - 1 + values().length) % values().length]` (someone should also do a more thorough test that those expressions work.

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+  }
+
+  public Arm(){
+    motorLencoder.setPositionConversionFactor(2*Math.PI/100);
```
Add a gearing constant

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+    Kd = SmartDashboard.getNumber("KD", Kd);
+    pid.setP(Kp);
+    pid.setI(Ki);
+    pid.setD(Kd);
```
All of these (Except maybe ArmENUM) should probably be moved to LiveWindow

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+    FFaccel = SmartDashboard.getNumber("FF: Acceleration", FFaccel);
+    goalPos = SmartDashboard.getNumber("GoalPosition", goalPos);
+    SmartDashboard.putNumber("ArmLencoderPos", motorLencoder.getPosition());
+    SmartDashboard.putString("ArmENUM", 
```
camelCase Enum

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+  public CANSparkMax motor = MotorControllerFactory.createSparkMax(Constants.Arm.port, TemperatureLimit.NEO);
+  public RelativeEncoder motorLencoder = motor.getEncoder();
+
+  public double encoderErrorTolerance = .1;
```
Document that this is Radians

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java:
```diff
@@ -4,25 +4,35 @@
 import edu.wpi.first.wpilibj2.command.Command;
 import edu.wpi.first.wpilibj2.command.Commands;
+import edu.wpi.first.wpilibj2.command.InstantCommand;
+import edu.wpi.first.wpilibj2.command.button.JoystickButton;
```
Run Organize Imports

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import com.revrobotics.SparkMaxPIDController;
+import com.revrobotics.CANSparkMax;
+import com.revrobotics.RelativeEncoder;
```
Run organize imports

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+  private double Kd = 1.2943;
+  private double Ki = 0.0;
+  
+  private double setpoint = 2.2;
```
Unused

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+    );
+    SmartDashboard.putNumber("KP", Kp);
+    SmartDashboard.putNumber("KI", Ki);
+    SmartDashboard.putNumber("KD", Kd);
```
See comment in `periodic()`

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+    SmartDashboard.putNumber("KP", Kp);
+    SmartDashboard.putNumber("KI", Ki);
+    SmartDashboard.putNumber("KD", Kd);
+    pid.setTolerance(2.5,10);
```
Magic numbers, make these into constants

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+    }
+    //help something went wrong
+    return null;
+  }
```
Just store the target `ArmPreset` and return that

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+    // because most people won't remember/want to do this long function chain
+    SmartDashboard.putString("Debooog", "CycleUp");
+    SmartDashboard.putNumber("GoalPosition", 
+    closeSnappedArmPos() != null ? closeSnappedArmPos().next().value : snappedArmPos().next().value
```
See comments in `periodic()` and constants

CoolSpy3 replied:
tab over

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+
+  public void cycleUp(){ 
+    // because most people won't remember/want to do this long function chain
+    SmartDashboard.putString("Debooog", "CycleUp");
```
Use a more descriptive name

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+    return null;
+  }
+
+  public void cycleUp(){ 
```
Make the method name more descriptive or add documentation

CoolSpy3 replied:
Forward call to `setPreset()`

CoolSpy3 replied:
Simply using the next and prev values could request that the arm move a large rotation after a single driver input. this may not always be desirable. Not requesting any changes, but should be kept in mind.

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+  }
+  
+  public void setPreset(ArmPreset preset){
+    SmartDashboard.putNumber("GoalPosition", preset.value);
```
See comments in `periodic()` and constants

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+    // if closeSnappedArmPos is working, swap based on it - otherwise use less accurate snapping
+  }
+  
+  public void cycleDown(){
```
See comments in cycleUp()

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+      //motor.set(dir * armFeed.calculate(goalPos + Math.PI/2, 0, 0));
+      motor.setVoltage(armFeed.calculate(goalPos, 0, 0)
+         + pid.calculate(motorLencoder.getPosition(), goalPos));
+    }
```
Make this into a default command

CoolSpy3 replied:
I thought about this, and actually, it makes more sense for our use case to have it as is. Please disregard this comment.

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+    } else {
+      //motor.set(dir * armFeed.calculate(goalPos + Math.PI/2, 0, 0));
+      motor.setVoltage(armFeed.calculate(goalPos, 0, 0)
+         + pid.calculate(motorLencoder.getPosition(), goalPos));
```
See thread in Mattermost

CoolSpy3 replied:
Consider using SparkMax onboard PIDController for faster adjustment

CoolSpy3 replied:
Velocity/Acceleration should change sign based on requested direction

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+    SmartDashboard.putNumber("KG", kG);
+    SmartDashboard.putString("Debooog", "No.");
+    SmartDashboard.putString("ArmENUM", 
+    snappedArmPos().toString()
```
tab over

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+    goalPos = SmartDashboard.getNumber("GoalPosition", goalPos);
+    SmartDashboard.putNumber("ArmLencoderPos", motorLencoder.getPosition());
+    SmartDashboard.putString("ArmENUM", 
+    (closeSnappedArmPos() != null) ? closeSnappedArmPos().toString() : snappedArmPos().toString()
```
tab over

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+package org.carlmontrobotics.robotcode2023.subsystems;
```
Check that all curly brackets are preceded by a space

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+      //motor.set(dir * armFeed.calculate(goalPos + Math.PI/2, 0, 0));
+      motor.setVoltage(armFeed.calculate(goalPos, 0, 0)
+         + pid.calculate(motorLencoder.getPosition(), goalPos));
+    }
```
This logic can be made more clear by using a single setVoltage call and a ternary operator around the feedforward args

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+  public double goalPos;
+
+  public enum ArmPreset {
+    INTAKE(0), MID(1), HIGH(2);
```
Some of these values are out of range for the arm. They should be set to the correct values, and the range should be documented.

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+  private ArmFeedforward armFeed = new ArmFeedforward(kS, kG, kV, kA);
+  private PIDController pid = new PIDController(Kp, Ki, Kd);
+  
+  public double goalPos;
```
Make private and add a getter

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java:
```diff
@@ -4,25 +4,35 @@
 
   public final Joystick driverController = new Joystick(0);
   public final Joystick manipulatorController = new Joystick(1);
   public final PowerDistribution pd = new PowerDistribution();
```
add newline

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+  private double FFvelocity = 6;
+  private double FFaccel = 6;
+  private ArmFeedforward armFeed = new ArmFeedforward(kS, kG, kV, kA);
+  private PIDController pid = new PIDController(Kp, Ki, Kd);
```
Make final

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+
+  public Arm(){
+    motorLencoder.setPositionConversionFactor(2*Math.PI/100);
+    motorLencoder.setPosition(-Math.PI / 2);
```
After changing to an absolute encoder, use setZeroOffset with a value defined in constants

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/Constants.java:
```diff
@@ -13,12 +15,23 @@
  */
 public final class Constants {
+
+    public static final class Arm{
```
Add space before `{`

CoolSpy3 commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,171 @@
+
+
+public class Arm extends SubsystemBase
+{
```
Move to previous line

CoolSpy3 reviewed (changes requested):
Here are my suggestions based on what I see rn. I recommend taking a look at these suggestions and then retesting to make sure the code still works.

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,101 @@
+    encoder.setPositionConversionFactor(1/15);
+    //encoder.setZeroOffset(-Math.PI / 2);
+    encoder.setPosition(-Math.PI / 2);
+    pid.setTolerance(2.5,10);
```
These numbers look wrong. Where did they come from?

FriedLongJohns replied:
PLaceholder values.
Fixed now, though.

CoolSpy3 committed:
arm subsystem bugfixes

CoolSpy3 committed:
code cleanup and bug fixes

CoolSpy3 committed:
enforce bounds on target positions

CoolSpy3 committed:
add SetArmWristPosition

CoolSpy3 committed:
fix logic bug

CoolSpy3 committed:
add isAtSetpoint methods

ProfessorAtomicManiac committed:
added values for the variables ryan put in

ProfessorAtomicManiac committed:
Added buttons, added tolerances to bounds of arm/wrist

ProfessorAtomicManiac committed:
Added joysticks for manual arm/wrist control

ProfessorAtomicManiac committed:
Working code that can drive arm to positions

CoolSpy3 committed:
use wildcard import

FriedLongJohns committed:
Nicer formatting + Fix SetARmWristPos behavior

always move wrist first, remove boolean
change 0 and 1's to arm and wrist constants

CoolSpy3 committed:
always stow wrist

CoolSpy3 committed:
Merge branch 'master' of https://github.com/DeepBlueRobotics/RobotCode2023 into Arm

CoolSpy3 committed:
Merge branch 'Arm' of https://github.com/DeepBlueRobotics/RobotCode2023 into HEAD

CoolSpy3 committed:
Revert "always stow wrist"

This reverts commit 7512c0ec479b9b640b0abfa45c04d45b8859682e.

ProfessorAtomicManiac committed:
Merge branch 'Arm' of https://github.com/DeepBlueRobotics/RobotCode2023 into ArmTest

ProfessorAtomicManiac committed:
bugfix

ProfessorAtomicManiac committed:
pid constants

ProfessorAtomicManiac committed:
Update src/main/java/org/carlmontrobotics/robotcode2023/commands/ArmPeriodic.java

Co-authored-by: Dean Brettle <dean@brettle.com>

ProfessorAtomicManiac committed:
Update src/main/java/org/carlmontrobotics/robotcode2023/commands/ArmPeriodic.java

Co-authored-by: Dean Brettle <dean@brettle.com>

ProfessorAtomicManiac committed:
resolved pr requests except for joysticks

ProfessorAtomicManiac committed:
Merge branch 'ArmTest' of https://github.com/DeepBlueRobotics/RobotCode2023 into ArmTest

sofiebudman committed:
added values for intake and outtake

ProfessorAtomicManiac committed:
Apply suggestions from code review

Co-authored-by: CoolSpy3 <coolspythree@gmail.com>

FriedLongJohns committed:
Nicer Constant formatting

ProfessorAtomicManiac committed:
Revert "Revert "always stow wrist""

This reverts commit e2e672536514995ab4b465e28c1d13954fb59e42.

ProfessorAtomicManiac committed:
Merge branch 'ArmTest' of https://github.com/DeepBlueRobotics/RobotCode2023 into ArmTest

sofiebudman committed:
added correct values to arm test

FriedLongJohns committed:
Move ALL SmartDashboard to LiveWindow

ProfessorAtomicManiac committed:
Implemented Trapezoidal control

ProfessorAtomicManiac committed:
Merge branch 'ArmTest' of https://github.com/DeepBlueRobotics/RobotCode2023 into ArmTest

FriedLongJohns committed:
fix getArm/WristGoal not being correct (?)

FriedLongJohns committed:
Merge branch 'Arm' into ArmTest

sofiebudman committed:
corrected wrist values

FriedLongJohns committed:
fix the not bug yes bug

CoolSpy3 committed:
Use inputModulus

FriedLongJohns committed:
https://github.com/DeepBlueRobotics/RobotCode2023/pull/6/#discussion_r1140757358

FriedLongJohns committed:
Fix OOB wrist/arm drive correction logic... maybe

FriedLongJohns committed:
import color

FriedLongJohns committed:
Correctly map controls as Uday begged

FriedLongJohns committed:
correct comment on flipping in/outtake

FriedLongJohns committed:
fix the logic... again

CoolSpy3 committed:
Merge pull request #6 from DeepBlueRobotics/ArmTest

Merging ArmTest into Arm branch

FriedLongJohns committed:
do SetARmWristPosition logic (crappily)

FriedLongJohns committed:
Merge pull request #7 from DeepBlueRobotics/ArmTest

do SetARmWristPosition logic (crappily)

ProfessorAtomicManiac committed:
added smartdashboard if initsendable no work

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/Constants.java:
```diff
@@ -33,37 +160,101 @@ public static final class Roller {
+          public static int BACK;
+          public static int START;
 
-            public static final int rollerIntakeConeButton = Button.kA.value;
-            public static final int rollerOuttakeConeButton = Button.kB.value;
-            public static final int rollerIntakeCubeButton = Button.kX.value;
-            public static final int rollerOuttakeCubeButton = Button.kY.value;
-            public static final int rollerStopButton = Button.kRightBumper.value;
```
Why is this removed?

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java:
```diff
@@ -23,25 +27,47 @@ public class RobotContainer {
       .onTrue(new RunRoller(roller, RollerMode.INTAKE_CONE, Constants.Roller.conePickupColor));
-    new JoystickButton(manipulatorController, Manipulator.rollerIntakeCubeButton)
-      .onTrue(new RunRoller(roller, RollerMode.INTAKE_CUBE, Constants.Roller.cubePickupColor));
-    new JoystickButton(manipulatorController, Manipulator.rollerOuttakeConeButton)
-      .onFalse(new RunRoller(roller, RollerMode.OUTTAKE_CONE, Constants.Roller.conePickupColor));
-    new JoystickButton(manipulatorController, Manipulator.rollerOuttakeCubeButton)
-      .onFalse(new RunRoller(roller, RollerMode.OUTTAKE_CUBE, Constants.Roller.cubePickupColor));
-    new JoystickButton(manipulatorController, Manipulator.rollerStopButton).onTrue(new InstantCommand(() -> roller.setSpeed(0)));
```
Why is this removed?

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java:
```diff
@@ -23,25 +27,47 @@ public class RobotContainer {
+    arm.setDefaultCommand(new ArmPeriodic(
+      arm, 
+      () -> inputProcessing(getStickValue(manipulatorController, Axis.kLeftY)),
+      () -> inputProcessing(getStickValue(manipulatorController, Axis.kRightX))
```
Is that supposed to be an X?

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java:
```diff
@@ -23,25 +27,47 @@ public class RobotContainer {
-    new JoystickButton(manipulatorController, Manipulator.rollerIntakeConeButton)
+    new JoystickButton(manipulatorController, Manipulator.toggleCube)
+      .onTrue(new InstantCommand(() -> {arm.object=0;}))
+      .onFalse(new InstantCommand(() -> {arm.object=1;}));
```
Use CONE and CUBE constants.

ProfessorAtomicManiac committed:
Finished OI except for shelf and substation

ProfessorAtomicManiac committed:
Removed magic number, used "asymmetric correction"

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java:
```diff
@@ -23,25 +27,47 @@ public class RobotContainer {
+      .onFalse(new InstantCommand(() -> {arm.object=1;}));
+    new JoystickButton(manipulatorController, Manipulator.toggleFront)
+      .onTrue(new InstantCommand(() -> arm.isFront=true))
+      .onFalse(new InstantCommand(() -> arm.isFront=false));
```
I think these will cause the `onTrue` commands to run when the button is pressed and the `onFalse` commands to run when the button is released. So a pess-and-release of the button will not cause the values to toggle.

ProfessorAtomicManiac replied:
They are not meant to be toggle but a "whenHeld" button where if a button is held then the mode switches

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java:
```diff
@@ -23,25 +27,47 @@ public class RobotContainer {
+      new SetArmWristPosition(arm.getArmGoal(GoalPos.HIGH), arm.getWristGoal(GoalPos.HIGH), arm)
+    );
+    new JoystickButton(manipulatorController, Manipulator.store).onTrue(
+      new SetArmWristPosition(arm.getArmGoal(GoalPos.STORED), arm.getWristGoal(GoalPos.STORED), arm)
```
Am I missing something or is this a duplicate of one of the earlier lines.

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/commands/ArmPeriodic.java:
```diff
@@ -0,0 +1,111 @@
+    TrapezoidProfile.Constraints armConstraints = new TrapezoidProfile.Constraints(MAX_FF_VEL[ARM], MAX_FF_ACCEL[ARM]);
+    var armProfile = new TrapezoidProfile(armConstraints,
+        new TrapezoidProfile.State(goals[0], goals[1]),
+        new TrapezoidProfile.State(currArmRad, armSubsystem.getArmPos()));
```
Isn't the second parameter supposed to be a velocity?

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/commands/ArmPeriodic.java:
```diff
@@ -0,0 +1,111 @@
+    TrapezoidProfile.Constraints wristConstraints = new TrapezoidProfile.Constraints(MAX_FF_VEL[WRIST], MAX_FF_ACCEL[WRIST]);
+    var wristProfile = new TrapezoidProfile(wristConstraints,
+        new TrapezoidProfile.State(goals[2], goals[3]),
+        new TrapezoidProfile.State(currWristRad, armSubsystem.getWristPos()));
```
Isn't the second parameter supposed to be a velocity?

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,369 @@
+        armMotor.setInverted(inverted[ARM]);
+        wristMotor.setInverted(inverted[WRIST]);
+
+        armEncoder.setPositionConversionFactor(rotationToRad);
+        wristEncoder.setPositionConversionFactor(rotationToRad);
```
What about the velocity conversion factors?

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/commands/ArmPeriodic.java:
```diff
@@ -0,0 +1,111 @@
+
+    if (Math.abs(goals[3]) < EPSILON) {
+      armSubsystem.setWristTarget(armSubsystem.getWristPos());
+    }
```
I don't think you want this. Imagine the arm is initially moving with a velocity of +1 rad/sec and the operator requests a speed of -1 rad/sec. Should something special happen when the speed is is temporarily near 0 along the way?

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/commands/ArmPeriodic.java:
```diff
@@ -0,0 +1,111 @@
+  }
+
+  // Copy and pasted from drivetrain, handles input from joysticks
+  public double[] getRequestedSpeeds(double deltaT) {
```
Currently this method is misnamed because it returns more than just speeds. But instead of renaming the method, consider moving the position calculations into `execute()` and making this method just return speeds. That would also mean you wouldn't need to pass in `deltaT`.

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/commands/SetArmWristPosition.java:
```diff
@@ -0,0 +1,48 @@
+        //need separate flipped case for front too since values are flipped when arm passes through bellypan
+        if (
+                ( arm.isFront && (arm.getArmPos() < arm.getArmGoal(GoalPos.MID)) && (arm.getArmPos() > armPos)) ||
+                (!arm.isFront && (arm.getArmPos() > arm.getArmGoal(GoalPos.MID)) && (arm.getArmPos() < armPos))
```
It's not entirely clear what `isFront` is supposed to represent, but based on the code in `Arm` and the fact that the default value for `isFront` is `true`, I think it's supposed to mean we are intaking in the front and outtaking in the back (i.e. the more common situation), not that the arm itself is "in front" of the verticals. If you want to know whether the arm is "in front", you should calculate that based on the arm position.

Next, note that if `isFront` is true, `arm.getArmGoal(GoalPos.MID)` will return a position in the *back*. This code seems to be assuming that it would be in the front.

In general, I recommend against using `arm.getArmGoal(GoalPos.MID)` as the position for making these decisions since further testing could cause us to change the position in order to score effectively. It would be better to create a new constant representing the angular distance from vertical below which you need to tuck the wrist before making the arm more vertical.

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/commands/SetArmWristPosition.java:
```diff
@@ -0,0 +1,48 @@
+                (!arm.isFront && (arm.getArmPos() > arm.getArmGoal(GoalPos.MID)) && (arm.getArmPos() < armPos))
+             ){
+            //but have special case where arm is in front and wrist would collide, so we have to move arm pos to mid before folding
+            if (arm.isFront){//TODO MAKE SURE THIS LINE IS CORRECT
```
Read [this other comment where I suggest a rewrite](https://github.com/DeepBlueRobotics/RobotCode2023/pull/1#discussion_r1140978722) first.

This could happen on either side.

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/commands/SetArmWristPosition.java:
```diff
@@ -0,0 +1,48 @@
+            //but have special case where arm is in front and wrist would collide, so we have to move arm pos to mid before folding
+            if (arm.isFront){//TODO MAKE SURE THIS LINE IS CORRECT
+                super.addCommands(
+                    new InstantCommand(() -> arm.setArmTarget(GoalPos.MID[0].armPos)),
```
Read [this other comment where I suggest a rewrite](https://github.com/DeepBlueRobotics/RobotCode2023/pull/1#discussion_r1140978722) first.

This needs to be changed to move to a clearance position on the same side the arm is already on.

Also, if arm/wrist is currently in CAD's "Move" named position (or otherwise positioned such the end of the rollers is inside the robot below the height of the bumpers), moving the arm before stowing the wrist will result in a collision.

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/commands/SetArmWristPosition.java:
```diff
@@ -0,0 +1,48 @@
+                new WaitUntilCommand(arm.wristPID::atSetpoint),//always stow first, avoid collision
+                new InstantCommand(() -> arm.setArmTarget(armPos)),
+                new WaitUntilCommand(arm.armPID::atSetpoint),
+                new InstantCommand(() -> arm.setWristTarget(wristPos)),//assume wristPos isn't gonna collide with drivetrain because that'd be a stupid thing for people to do and I don't wanna add more if statements
```
Read [this other comment where I suggest a rewrite](https://github.com/DeepBlueRobotics/RobotCode2023/pull/1#discussion_r1140978722) first.

While I sympathize, people do stupid things all the time and we can't afford for them to result in a broken robot.

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/commands/SetArmWristPosition.java:
```diff
@@ -0,0 +1,48 @@
+                new InstantCommand(() -> arm.setArmTarget(armPos)),
+                new InstantCommand(() -> arm.setWristTarget(wristPos)),
+                new WaitUntilCommand(() -> {return (arm.wristPID.atSetpoint() && arm.armPID.atSetpoint());})
+            );
```
Read [this other comment where I suggest a rewrite](https://github.com/DeepBlueRobotics/RobotCode2023/pull/1#discussion_r1140978722) first.

This is a good plan *if* the arm is high enough to be clear *and* it is moving *higher* (i.e. away from vertical). However, if it is moving *lower* you could collide. Note that with your current logic, you will reach this code in that case.

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/commands/SetArmWristPosition.java:
```diff
@@ -0,0 +1,48 @@
+import edu.wpi.first.wpilibj2.command.SequentialCommandGroup;
+import edu.wpi.first.wpilibj2.command.WaitUntilCommand;
+
+public class SetArmWristPosition extends SequentialCommandGroup {
```
This probably needs a rewrite. Consider approaching the problem as follows:

1. Which of the following statements is true? "From any position there is always a direction I can rotate the arm joint without colliding" or "From any position there is always a direction I can rotate the wrist joint without colliding". Use the answer to your advantage by always rotating that joint in the available direction first.
2. To determine which direction to rotate it, draw a side view of the robot (including bumpers) and consider the various spots that the end of the roller could be. Can you find a relatively simple rule for determining which direction to rotate given where the end of the roller is?
3. Rotate the joint from step 1 in the direction from step 2 at least enough to allow you to rotate the other joint without colliding. Note that this doesn't necessarily mean that the first joint will always be rotated to the same position.
4. Rotate the second joint to a position that allows you full freedom in the rotation of the first joint.
5. Now to move to the target arm/wrist position from this position without colliding imagine you were *starting* at the target position followed steps 2-4 determine how to "get free". Do the resulting movements in reverse to get from your free position to the target position.

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,369 @@
+    public PIDController armPID = new PIDController(kP[ARM], kI[ARM], kD[ARM]);
+    public PIDController wristPID = new PIDController(kP[WRIST], kI[WRIST], kD[WRIST]);
+
+    private SendableBuilder senb;
```
I don't think you need this.

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,369 @@
+        wristPID.setTolerance(posToleranceRad[WRIST], velToleranceRadPSec[WRIST]);
+
+        SmartDashboard.putBoolean("Toggle", false);
+        senb.update();
```
This should not be needed.

brettle replied:
Instead, call `SmartDashboard.putData("Arm", this)`.

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,369 @@
+        armPID.setTolerance(posToleranceRad[ARM], velToleranceRadPSec[ARM]);
+        wristPID.setTolerance(posToleranceRad[WRIST], velToleranceRadPSec[WRIST]);
+
+        SmartDashboard.putBoolean("Toggle", false);
```
Not clear what this is for.

Note that you probably want an "Apply Risky Changes" button that, when clicked, sets the values that you don't want set with every keystroke in Shuffleboard. The setters for those properties would set an intermediate variable and the "Apply Risky Changes" would copy the values from those variables into the real variables.

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,369 @@
+
+    @Override
+    public void periodic() {
+        // TODO: REMOVE THIS WHEN PID CONSTANTS ARE DONE
+        armPID.setP(kP[ARM]);
+        armPID.setI(kI[ARM]);
+        armPID.setD(kD[ARM]);
+        wristPID.setP(kP[WRIST]);
+        wristPID.setI(kI[WRIST]);
+        wristPID.setD(kD[WRIST]);
+        if (SmartDashboard.getBoolean("Toggle", false)) {
+            senb.update();
+        }
```
None of this should be needed.

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,369 @@
+        }
+    }
+
+    public void driveArm(double vel, double accel) {
+        double kgv = getKg();
+        double armFeedVolts = kgv * getCoM().getAngle().getCos() + armFeed.calculate(vel, accel);
+        double armPIDVolts = armPID.calculate(armEncoder.getPosition(), goalPosRad[ARM]);
+        if ((getArmPos() > ARM_UPPER_LIMIT_RAD && vel > 0) || 
+            (getArmPos() < ARM_LOWER_LIMIT_RAD && vel < 0)) {
+            armFeedVolts = kgv * getCoM().getAngle().getCos() + armFeed.calculate(0, 0);
+        }
+        // TODO: REMOVE WHEN DONE WITH TESTING (ANY CODE REVIEWERS, PLEASE REJECT MERGES
+        // TO MASTER IF THIS IS STILL HERE)
+        SmartDashboard.putNumber("ArmFeedVolts", armFeedVolts);
+        SmartDashboard.putNumber("ArmPIDVolts", armPIDVolts);
+        double volts = armFeedVolts + armPIDVolts;
+        SmartDashboard.putNumber("ArmTotalVolts", volts);
+        armMotor.setVoltage(volts);
+    }
+
+    public void driveWrist(double vel, double accel) {
+        double kgv = wristFeed.calculate(getWristPosRelativeToGround(), 0, 0);
+        double wristFeedVolts = wristFeed.calculate(getWristPosRelativeToGround(), vel, accel);
+        double wristPIDVolts = wristPID.calculate(getWristPos(), goalPosRad[WRIST]);
+        if ((getWristPos() > WRIST_UPPER_LIMIT_RAD && vel > 0) || 
+            (getWristPos() < WRIST_LOWER_LIMIT_RAD && vel < 0)) {
+            wristFeedVolts = kgv;
+        }
+        // TODO: REMOVE WHEN DONE WITH TESTING (ANY CODE REVIEWERS, PLEASE REJECT MERGES
+        // TO MASTER IF THIS IS STILL HERE)
+        SmartDashboard.putNumber("WristFeedVolts", wristFeedVolts);
+        SmartDashboard.putNumber("WristPIDVolts", wristPIDVolts);
+        double volts = wristFeedVolts + wristPIDVolts;
+        SmartDashboard.putNumber("WristTotalVolts", volts);
+        wristMotor.setVoltage(volts);
+    }
```
These are no longer used.

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,369 @@
+    public void driveArm(TrapezoidProfile.State profile) {
+        double kgv = getKg();
+        double armFeedVolts = kgv * getCoM().getAngle().getCos() + armFeed.calculate(profile.velocity, 0);
+        double armPIDVolts = armPID.calculate(armEncoder.getPosition(), profile.position);
```
`armEncoder.getPosition()` does not have the appropriate range.

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,369 @@
+    public void driveWrist(TrapezoidProfile.State profile) {
+        double kgv = wristFeed.calculate(getWristPosRelativeToGround(), 0, 0);
+        double wristFeedVolts = kgv * getCoM().getAngle().getCos() + wristFeed.calculate(profile.velocity, 0);
+        double wristPIDVolts = wristPID.calculate(wristEncoder.getPosition(), profile.position);
```
`wristEncoder.getPosition()` does not have the appropriate range.

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,369 @@
+        builder.addDoubleProperty("WristGoalPos", () -> goalPosRad[WRIST], goal -> goalPosRad[WRIST] = goal);
+        builder.addDoubleProperty("ArmGoalPos",   () -> goalPosRad[ARM], goal -> goalPosRad[ARM] = goal);
+        builder.addDoubleProperty("ArmPos",       () -> getArmPos(), null);
+        builder.addDoubleProperty("WisPos",       () -> getWristPos(), null);
+        builder.addDoubleProperty("RobotWisPos",  () -> getWristPosRelativeToGround(), null);
```
Wis -> Wrist

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,369 @@
+        if (isFront) {
+            return pos[object].armPos;
+        } else {
+            return Rotation2d.fromRadians(-Math.PI).minus(Rotation2d.fromRadians(pos[object].armPos)).plus(Rotation2d.fromDegrees(10)).getRadians();
```
Positions on the opposite side can be unrelated to positions due to wrist asymmetry.

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,369 @@
+        if (isFront) {
+            return pos[object].wristPos;
+        } else {
+            return Rotation2d.fromRadians(pos[object].wristPos).unaryMinus().plus(Rotation2d.fromDegrees(10)).getRadians();
```
Positions on the opposite side can be unrelated to positions due to wrist asymmetry.

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/Constants.java:
```diff
@@ -100,24 +99,23 @@ public static final class Arm {
+        // if intake/outtake on back, the "negative" pos will be used
         // 0 = CUBE, 1 = CONE
-        // {Cube, Cone}
+        // {Cone, Cube}
```
These 2 lines contradict each other.

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/Constants.java:
```diff
@@ -132,9 +130,12 @@ public static final class GoalPos {
+          new GoalPos(Units.degreesToRadians(-72.5), Units.degreesToRadians(92.21)),
+          new GoalPos(Units.degreesToRadians(-66.6), Units.degreesToRadians(48.78)) 
         };
+        // Rollers are asymmetric, so when moving rollers to other side of robot, this value will correct
+        // the roller so that the rollers will act as if its on the front side
+        public static double asymmetricCorrectionDeg = 10;
```
Unfortunately, it's not that simple. Consider the case of placing a cube in the front instead of the back.

ProfessorAtomicManiac committed:
Fixed Trapezoidal Control

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/commands/ArmPeriodic.java:
```diff
@@ -0,0 +1,105 @@
+
+  // Called every time the scheduler runs while the command is scheduled.
+  @Override
+  public void execute() {
```
Some of what is in this method actually belongs in `Arm.periodic()`. As it stands, goals set by other commands will be ignored. Basically, what you want is for this method to just calculate the goal profile states based on the requested speeds, and then pass those goal profile states to the `Arm` subsystem.  Then `Arm.periodic()` should use whatever the the goal profile states are to create and calculate the profiles and drive the arm/wrist (i.e. the code which does that in this method should be moved to `Arm.periodic()`).

ProfessorAtomicManiac committed:
Added back and front pos (VALUES NOT PUT IN YET)

ProfessorAtomicManiac committed:
forgot to switch butons

TuskAct2 committed:
made requested changes by dean

DriverStationComputer committed:
change smartdashboard

TuskAct2 committed:
added new back angles for low - high pos

TuskAct2 committed:
Merge branch 'Arm' of https://github.com/DeepBlueRobotics/RobotCode2023 into Arm

ProfessorAtomicManiac committed:
moved code driving motor to positions into periodic

ProfessorAtomicManiac committed:
Merge branch 'Arm' of https://github.com/DeepBlueRobotics/RobotCode2023 into Arm

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -194,17 +196,14 @@ public double getWristPosRelativeToGround() {
-        goalPosRad[WRIST] = getWristClampedGoal(target);
+    public void setArmTarget(double targetPos, double targetVel) {
+        goalState[ARM].position = getArmClampedGoal(targetPos);
+        goalState[ARM].velocity = MathUtil.clamp(targetVel, 0, MAX_FF_VEL[ARM]);
```
This is wrong and not needed. Wrong because the goal velocity could be negative and not needed because the TrapezoidProfile velocity limit will take care of it.

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -194,17 +196,14 @@ public double getWristPosRelativeToGround() {
-        setWristTarget(targetWrist);
+    public void setWristTarget(double targetPos, double targetVel) {
+        goalState[WRIST].position = getWristClampedGoal(targetPos);
+        goalState[WRIST].velocity = MathUtil.clamp(targetVel, 0, MAX_FF_VEL[WRIST]);
```
This is wrong and not needed. Wrong because the goal velocity could be negative and not needed because the TrapezoidProfile velocity limit will take care of it.

ProfessorAtomicManiac committed:
Moved Trapezoidal constraints into Arm Subsystem

DriverStationComputer committed:
added front angles for arm and wrist.

DriverStationComputer committed:
Merge branch 'Arm' of https://github.com/DeepBlueRobotics/RobotCode2023 into Arm

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,350 @@
+    @Override
+    public void periodic() {
+        double currTime = Timer.getFPGATimestamp();
+        double deltaT = currTime - lastTime;
```
A bug is causing deltaT to keep growing. Do you see it?

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,350 @@
+        wristEncoder.setVelocityConversionFactor(rotationToRad);
+
+        wristEncoder.setZeroOffset(offsetRad[ARM]);
+        armEncoder.setZeroOffset(offsetRad[WRIST]);
```
There are bugs on these 2 lines.

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,350 @@
+    private void driveArm(TrapezoidProfile.State state) {
+        double kgv = getKg();
+        double armFeedVolts = kgv * getCoM().getAngle().getCos() + armFeed.calculate(state.velocity, 0);
+        double armPIDVolts = armPID.calculate(armEncoder.getPosition(), state.position);
```
armEncoder.getPosition() will not be in the correct range.

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:
```diff
@@ -0,0 +1,350 @@
+        wristConstraints = new TrapezoidProfile.Constraints(MAX_FF_VEL[WRIST], MAX_FF_ACCEL[WRIST]);
+        var wristProfile = new TrapezoidProfile(wristConstraints,
+            goalState[WRIST],
+            new TrapezoidProfile.State(getWristPos(), getWristVel()));
```
keep track of the current setpoints and create profiles
that start at the current setpoints (instead of the current positions).

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/commands/ArmPeriodic.java:
```diff
@@ -0,0 +1,86 @@
+    double deltaT = currTime - lastTime;
+
+    double goalArmRad = armSubsystem.getArmPos() + speeds[ARM] * deltaT;
+    double goalWristRad = armSubsystem.getWristPos() + speeds[WRIST] * deltaT;
```
Adjust the arm/wrist target positions based on the current target positions instead of current actual positions.

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/commands/ArmPeriodic.java:
```diff
@@ -0,0 +1,86 @@
+
+    // Retrieve the profiled setpoint for the next timestep. This setpoint moves
+    // toward the goal while obeying the constraints.
+    goalArmRad = armSubsystem.getArmClampedGoal(goalArmRad);
+    goalWristRad = armSubsystem.getWristClampedGoal(goalWristRad);
```
These lines aren't needed because setArmTarget() is responsible for clamping.

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/commands/ArmPeriodic.java:
```diff
@@ -0,0 +1,86 @@
+    // toward the goal while obeying the constraints.
+    goalArmRad = armSubsystem.getArmClampedGoal(goalArmRad);
+    goalWristRad = armSubsystem.getWristClampedGoal(goalWristRad);
+    armSubsystem.setArmTarget(goalArmRad, speeds[ARM]);
+    armSubsystem.setWristTarget(goalWristRad, speeds[WRIST]);
```
Don't update target target speeds. Just leave them at whatever the current target speeds are.

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/Constants.java:
```diff
@@ -18,6 +21,177 @@
+
+        // TODO: Determine actual max vel/accel
+        public static double[] MAX_FF_VEL = {1, 1}; // rad / s
+        public static double[] MAX_FF_ACCEL = {1, 1}; // rad / s
```
Comment has wrong units.

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/Constants.java:
```diff
@@ -18,6 +21,177 @@
+        // PID
+        // FIXME BOTH WRIST AND ARM NEED TO TEST PID (Wrist PID never tested)
+        // Arm, Wrist
+        public static double[] kP = {6.7868, 4.7391}; // 4.2736 for arm from sysid was tested and it worked fine (V / rad)
+        public static double[] kI = {0, 0}; // (V / (rad * s) )
+        public static double[] kD = {4.4327, 0.69517}; // 0 for arm from sysid was tested and it worked fine (V / (rad / s) )
```
Why aren't we using the values from sysid? Were the new values arrived at via empirical tuning?

brettle commented on src/main/java/org/carlmontrobotics/robotcode2023/Constants.java:
```diff
@@ -18,6 +21,177 @@
+        public static double[] velToleranceRadPSec = { 0.5, 0.5 }; // rad/s
+
+        // rad, rad/s
+        public static TrapezoidProfile.State[] goalState = { new TrapezoidProfile.State(-Math.PI / 2, 0), new TrapezoidProfile.State(0, 0) }; // rad
```
This isn't a constant. It should be a private member of Arm.java and very importantly it should be initialized to the current position at startup so that the arm/wrist don't try to move unless we tell them to.

CoolSpy3 committed:
fix trapezoidprofile implementation

CoolSpy3 committed:
fix PR comments

CoolSpy3 committed:
ArmTeleop bugfixes

CoolSpy3 committed:
add OI

CoolSpy3 committed:
update wrist offset

CoolSpy3 committed:
debugging

DriverStationComputer committed:
change values

DriverStationComputer committed:
Merge branch 'Arm' of https://github.com/DeepBlueRobotics/RobotCode2023 into Arm

CoolSpy3 committed:
cleanup

CoolSpy3 committed:
reset goal on enable

CoolSpy3 committed:
bugfix

DriverStationComputer committed:
debugging changes

TuskAct2 committed:
setposition command

TuskAct2 committed:
Merge branch 'Arm' of https://github.com/DeepBlueRobotics/RobotCode2023 into Arm

CoolSpy3 committed:
Merge branch 'master' into Arm

CoolSpy3 committed:
add missing changes from merge
